src/services/completions.ts

```diff
@@ -1280,33 +1281,54 @@ namespace ts.Completions {
                     // This is just to avoid adding duplicate completion entries.
                     //
                     // If `symbol.parent !== ...`, this is an `export * from "foo"` re-export. Those don't create new symbols.
-                    if (typeChecker.getMergedSymbol(symbol.parent!) !== resolvedModuleSymbol
-                        || some(symbol.declarations, d =>
-                            // If `!!d.name.originalKeywordKind`, this is `export { _break as break };` -- skip this and prefer the keyword completion.
-                            // If `!!d.parent.parent.moduleSpecifier`, this is `export { foo } from "foo"` re-export, which creates a new symbol (thus isn't caught by the first check).
-                            isExportSpecifier(d) && (d.propertyName ? isIdentifierANonContextualKeyword(d.name) : !!d.parent.parent.moduleSpecifier))) {
+                    if (typeChecker.getMergedSymbol(symbol.parent!) !== resolvedModuleSymbol) {
                         continue;
                     }
-
-                    const isDefaultExport = symbol.escapedName === InternalSymbolName.Default;
-                    if (isDefaultExport) {
-                        symbol = getLocalSymbolForExportDefault(symbol) || symbol;
+                    // If `!!d.name.originalKeywordKind`, this is `export { _break as break };` -- skip this and prefer the keyword completion.
+                    if (some(symbol.declarations, d => isExportSpecifier(d) && !!d.propertyName && isIdentifierANonContextualKeyword(d.name))) {
+                        continue;
                     }
-
-                    const origin: SymbolOriginInfoExport = { kind: SymbolOriginInfoKind.Export, moduleSymbol, isDefaultExport };
-                    if (detailsEntryId || stringContainsCharactersInOrder(getSymbolName(symbol, origin, target).toLowerCase(), tokenTextLowerCase)) {
-                        symbols.push(symbol);
-                        symbolToSortTextMap[getSymbolId(symbol)] = SortText.AutoImportSuggestions;
-                        symbolToOriginInfoMap[getSymbolId(symbol)] = origin;
+                    // If `!!d.parent.parent.moduleSpecifier`, this is `export { foo } from "foo"` re-export, which creates a new symbol (thus isn't caught by the first check).
+                    if (some(symbol.declarations, d => isExportSpecifier(d) && !d.propertyName && !!d.parent.parent.moduleSpecifier)) {
+                        // If we haven’t yet seen the original symbol, note the alias so we can add it at the end if the original doesn’t show up eventually
+                        const nearestExportLocalSymbol = forEachAlias(typeChecker, symbol, alias => forEach(alias.declarations, d => d.localSymbol));
+                        if (nearestExportLocalSymbol && !symbolToOriginInfoMap[getSymbolId(nearestExportLocalSymbol)]) {

```

Comment:
This whole area was trickier than I expected, and I’m not 100% confident that this is totally sound, or the most efficient way to do this. The re-export logic had to change because now that some filtering is potentially happening a step before this runs, it’s possible that we’ll see a re-export but never see the original, in which case we _shouldn’t_ filter out the re-export. So, if we see a re-export, we try to determine if we’ve already seen the thing we’re re-exporting. I’m doing this semantically, but maybe it’s cheaper and just as good to do it syntactically? I was afraid of missing corner cases that way.

Reply:
Update: this _is_ currently wrong (that’s why tests are failing)—although I’m pretty sure some iteration of this logic worked yesterday—but it’s quite difficult to discern exactly what kind of symbol appears in the `symbols` array here; it seems like sometimes it’s the `localSymbol` and sometimes it’s the `exportSymbol`. Still investigating and revising this section.